packages/menu/src/Menu.ts

```diff
@@ -201,30 +370,90 @@ export class Menu extends SpectrumElement {
             item = this.menuItems[index] as MenuItem;
         }
         index = Math.max(index, 0);
+        const selectedItemsMap = new Map<MenuItem, boolean>();
+        const selected: string[] = [];
+        const selectedItems: MenuItem[] = [];
         this.menuItems.forEach((item, i) => {
+            if (item.selected) {
+                selectedItemsMap.set(item, true);
+                selected.push(item.value);
+                selectedItems.push(item);
+            }
             if (i !== index) {
                 item.focused = false;
             }
         });
+        this.selectedItemsMap = selectedItemsMap;
+        this.selected = selected;
+        this.selectedItems = selectedItems;
         this.focusedItemIndex = index;
         this.focusInItemIndex = index;
     }
 
-    private prepItems = (): void => {
-        this.menuItems = [
-            ...this.querySelectorAll(`[role="${this.childRole}"]`),
-        ] as MenuItem[];
-        if (!this.menuItems || this.menuItems.length === 0) {
-            return;
+    // debounce update this.menuItems so they're in DOM order
+    private async updateMenuItems(): Promise<void> {
+        this.itemsChanged = true;
+
+        return new Promise((resolve) => {
+            // debounce so we only actually update once
+            requestAnimationFrame(() => {
+                if (!this.itemsChanged) {
+                    resolve();
+                }
+
+                this.menuItems = [];
+
+                const slotElements =
+                    this.menuSlot.assignedElements({ flatten: true }) || [];
+                for (const slotElement of slotElements) {
+                    const childItems: MenuItem[] =
+                        slotElement instanceof MenuItem
+                            ? [slotElement as MenuItem]
+                            : ([
+                                  ...slotElement.querySelectorAll(
+                                      `[role="${this.childRole}"]`
+                                  ),
+                              ] as MenuItem[]);
+                    for (const childItem of childItems) {
+                        if (this.menuItemSet.has(childItem)) {
+                            this.menuItems.push(childItem);
```

Comment:
Related to the below comment around who owns `this.menuItems` to manage keyboard and focus work, it seems like `this.menuItems` is a larger collection that `this.menuItemSet` (which is about selection, right?) in way that questions this logic...

Reply:
thought I'd responded yesterday. With the approach here, `this.menuItemSet` is used to manage *what* items are owned, while `this.menuItems` has thee same items in the right DOM order.

I can't say I like it, and one thing I'm thinking is that moving more of the logic here into a `menuItems` getter (along with an invalidation flag), might help clean this up, and address some of the other feedback you have here.

Reply:
Let me know what you think of the way this is re-worked now.